Add support for Key-Value data in log records.

tmccombs opened the pull request:
See https://github.com/rust-lang-nursery/log/issues/328

tmccombs commented on src/fmt/mod.rs:
```diff
@@ -289,6 +294,28 @@ impl<'a> DefaultFormat<'a> {
+            return Ok(())
+        }
+        let kvs = record.key_values();
+        if !self.written_header_value && kvs.count() > 0 {
```
I don't love this, but I couldn't think of a way of doing it I liked any better.

tmccombs commented on src/fmt/mod.rs:
```diff
@@ -289,6 +294,28 @@ impl<'a> DefaultFormat<'a> {
+            let open_brace = self.subtle_style("[");
+            write!(self.buf, "{}", open_brace)?;
+        }
+        // Ideally we would be able to transport the original io::Error through the visit
```
See https://github.com/rust-lang-nursery/log/pull/336

simao commented:
Any news on this? Could we at least merge it behind a feature flag or something? Does this need any changes besides a rebase?

hendrikmaus commented:
I would also love to start using it.

One good addition to the pull-request would probably be a usage example.

simao commented:
I tried these changes and they work well. As far as usage examples go, I think that is done in the `log` crate? https://docs.rs/log/0.4.16/log/#structured-logging

Users of env_log would just need to use the `log/kv_unstable` feature, which I suppose should be mentioned somewhere?

One thing I noticed is maybe the target should be written after the kv pairs. Currently we get:

```
[2022-04-01T08:52:43Z INFO  tide::server] Server listening on http://0.0.0.0:8080
[2022-04-01T08:52:50Z INFO  method=POST path=/sensors/inside_weather status=200 - OK duration=11.487824ms tide::log::middleware] --> Response sent
```

But IMO the following looks cleaner, since target stays aligned with lines that do not have kv pairs:

```
[2022-04-01T08:54:12Z INFO  tide::log::middleware method=POST path=/sensors/inside_weather status=200 - OK duration=10.166265ms] --> Response sent
```

Also wanted to mentioned that it might be worth it to consider putting the structured data in different lines, to improve readability. At least for development, IMO the following looks better:

```
[2022-04-01T08:57:16Z INFO  tide::server] Server listening on http://0.0.0.0:8080
[2022-04-01T08:57:19Z INFO  tide::log::middleware
    method POST
    path /sensors/inside_weather] <-- Request received
[2022-04-01T08:57:19Z INFO  tide::log::middleware
    method POST
    path /sensors/inside_weather
    status 200 - OK
    duration 10.842409ms] --> Response sent
```

This is what `femme` does, and it's just using `write!(self.0, "\n    {} {}", key, value)` in `visit_pairs`. 

Though in production we'd probably want everything on the same line.

hendrikmaus commented:
Thanks for pointing to the examples in the log crate. That makes sense. 

As to your other points: I'd second both of them.

svanharmelen commented:
I was just about to open an issue when I found this PR... It would be awesome if this could be merged so I can keep using `env_logger` and start using structured logging!

svanharmelen commented:
Thanks for the rebase @tmccombs! After using this a little I'm thinking that it would (maybe) be good to make it configurable where the K/V's are written? As I'm used that the are written behind the actual log message instead of in front of it.

When reading a log of log files that all log a different number of K/V's, you really have to scan al lot for where the actual message is. Where if the K/V's are added at the end, all the messages are directly above each other 🤷🏻‍♂️

tmccombs commented:
I added a commit that moves the Key values behind the rest of the option, and made it configurable whether it is on a single line or on separate lines, as @simao suggested.

svanharmelen commented:
Looks really good!! Thanks for your efforts, I'll be using your branch until this get (hopefully) merged 👍🏻

fgimian commented:
I suggest also checking out frameworks like [zerolog](https://github.com/rs/zerolog) which present structured log fields as `key=value` fields on the same line and with pretty colours.

![image](https://user-images.githubusercontent.com/1811813/236113893-431a8bbc-ff99-4b2d-b0ce-1f903ad734bd.png)

I think output similar to this would be nicer personally 😄

prasannavl commented:
Any reason on what this is blocked on? Would be great to get this merged, even until an experimental flag.

epage commented:
@tmccombs at this point my thoughts are
- Mirror what tracing-subscriber does for formatting and control of it
- Rename the feature to `unstable-kv`, other features are using dashes at this point and I tend to put `unstable-` first.

epage commented:
Hadn't looked into that before.  Thanks for listing those considerations.  As this is unstable, let's focus on an MVP and just use the formatting they do by default and not provide any configuration.  We can then explore in an Issue what our formatting configuration options are, both those from the two PRs and from tracing-subscriber.

epage commented on src/fmt/kv.rs:
```diff
@@ -0,0 +1,45 @@
+///
+/// This trait determines how key/value pairs for structured logs are serialized within the default
+/// format.
+pub(crate) type KvFormatFn = dyn Fn(&mut Formatter, &dyn Source) -> io::Result<()> + Sync + Send;
```
Is there a reason you didn't match `FormatFn` in making this `Box`ed?

tmccombs replied:
Because I use it as a `&` [here](https://github.com/rust-cli/env_logger/pull/137/files#diff-3d324710ba49d50da6915fa5a9bee1c6fbd2a8a97a7d8aa45730c3f9a71f54baR284). But if you would prefer fo rme to make it boxed, I'm fine with that too.

epage replied:
Prefer consistency but not strong enough

epage commented:
`KvFormatFn` also works and fits within the style of the existing API.  Thanks!

tmccombs commented:
I've resolved conflicts and squashed the commits.

epage commented on src/fmt/mod.rs:
```diff
@@ -70,6 +91,38 @@ pub enum TimestampPrecision {
+/// The values can be suppressed, included in a trailer at the end of the line, or
+/// included as multiple lines after the message.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum KVStyle {
```
Is this a zombie from resolving the merge conflicts?

tmccombs replied:
ah, yes it is

epage commented on src/fmt/mod.rs:
```diff
@@ -41,13 +58,17 @@ use log::Record;
 
 #[cfg(feature = "humantime")]
 pub use self::humantime::Timestamp;
+#[cfg(feature = "unstable-kv")]
+pub use self::kv::*;
```
If you are already going back to make a change, can you you check this on `RUSTDOCFLAGS=--cfg=docsrs cargo +nightly doc --all-features` (I'm hoping that is correct).

Each API item behind a `cfg` should have a note in the documentation, like [`clap::command` requiring `cargo`](https://docs.rs/clap/latest/clap/macro.command.html).

Most likely, you'll need to either duplicate this `cfg` on the items or move the `cfg` into the `kv` mod.

tmccombs replied:
Looks like it works as is:

![image](https://github.com/rust-cli/env_logger/assets/2541726/4eed76a1-4bfd-4f38-a7b3-5c9d46f8586c)

![image](https://github.com/rust-cli/env_logger/assets/2541726/7ec8a794-d53b-4fa7-93b0-dc6d21d6c84c)

KodrAus commented:
Hey all! :wave:

Firstly, thank you for picking up and carrying this project forwards 🙏

We're (finally) about to stabilize the key-value support in `log` (see https://github.com/rust-lang/log/pull/617), making this possible to support in `env_logger` without needing to fiddle with any unstable feature flags. If you'd like any help finishing off support here please do let me know.

tmccombs commented:
I've updated to use the stable version of the log kv feature. Fixed merge conflicts, and fixed the compilation error when the color option is not enabled.
